### PR TITLE
Data Layer Setup

### PR DESCRIPTION
**Pull request by Vanshbajaj**
1. Steup Data Layer
2. Add All Things That Required for Data Layer 
3. Github Action will run on the PR only
4. Updated to KSP

### CODE REVIEW

File: data/src/main/java/com/practical/data/network/DataModule.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.apollographql.apollo3.ApolloClient
+import com.practical.data.repository.CharacterRepositoryImpl
+import com.practical.domain.repository.CharacterRepository
+import com.practical.domain.usecases.GetCharactersUseCase
+import dagger.Module
+import dagger.Provides
+import javax.inject.Singleton
+
+@Module
+object DataModule {
```
Comment (shubhamsinghmutualmobile) [inline data/src/main/java/com/practical/data/network/DataModule.kt:12]: No `@InstallIn`?

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [versions]
 agp = "8.6.0"
 converterScalars = "2.9.0"
-dagger = "2.51.1"
+dagger = "2.44"
```
Comment (shubhamsinghmutualmobile) [inline gradle/libs.versions.toml:4]: Why downgrade?

**Review by shubhamsinghmutualmobile — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -6,8 +6,12 @@ on:
       - '*'   # Trigger on pushes to the feature branch
   pull_request:
     branches:
+
+      - '*'  # Trigger on PRs targeting the feature branch
+
```
Comment (shreyansh-mm) [inline .github/workflows/ci.yml:11]: Why do we need it twice?

File: data/build.gradle.kts [hunk 4]
```diff
@@ -1,6 +1,8 @@
 plugins {
     alias(libs.plugins.android.library)
     alias(libs.plugins.kotlin.android)
+    id("kotlin-kapt")
```
Comment (shreyansh-mm) [inline data/build.gradle.kts:4]: We can use KSP with upgraded version of dagger (2.48 and above)

File: data/src/main/java/com/practical/data/network/DataModule.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    @Provides
+    @Singleton
+    fun provideApolloClient(): ApolloClient {
+        return ApolloClient.Builder()
+            .serverUrl("https://rickandmortyapi.com/graphql")
```
Comment (shreyansh-mm) [inline data/src/main/java/com/practical/data/network/DataModule.kt:17]: Let's move the URL to a constants file for now

File: domain/src/main/java/com/practical/domain/CharacterDataClass.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.practical.domain
+
+
+
+data class CharacterDataClass(
```
Comment (shreyansh-mm) [inline domain/src/main/java/com/practical/domain/CharacterDataClass.kt:5]: formatting

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/practical/data/network/DataModule.kt (hunk 1 above)
Comment (Vanshbajaj) [reply data/src/main/java/com/practical/data/network/DataModule.kt:12]: Using Dagger2

File: gradle/libs.versions.toml (hunk 2 above)
Comment (Vanshbajaj) [reply gradle/libs.versions.toml:4]: Dagger 2 Was not working so i need downgrade, soon updating to ksp

File: data/src/main/java/com/practical/data/network/DataModule.kt (hunk 5 above)
Comment (Vanshbajaj) [reply data/src/main/java/com/practical/data/network/DataModule.kt:17]: Sure

File: data/build.gradle.kts (hunk 4 above)
Comment (Vanshbajaj) [reply data/build.gradle.kts:4]: Sure

File: build.gradle.kts [hunk 7]
```diff
@@ -4,6 +4,8 @@ plugins {
     alias(libs.plugins.android.library) apply false
     alias(libs.plugins.apollo.android) apply false
     alias(libs.plugins.detekt.plugin) apply true  // Apply Detekt globally
+   alias(libs.plugins.ksp.plugin) apply  false
```
Comment (shreyansh-mm) [inline build.gradle.kts:7]: alignment

**Review by shreyansh-mm — APPROVED** [review]

**Pull request merged**